add rough docs for unevaluatedItems (#205)
json-schema-org/understanding-json-schema: source/reference/array.rst

    // The schema fails validation because ``"unevaluatedItems": false"``
    // specifies no extra values should exist.

``unevaluatedItems`` is used mainly when extending a tuple. So with
this in mind, you can make a "half-closed" schema: something useful
when you want to keep the first two arguments, but also add more in
certain situations. ("Closed" to two arguments in some places, "open"
to more arguments when you need it to be.)

.. schema_example::

    {
      "$id": "https://example.com/my-tuple",

      "type": "array",
      "prefixItems": [
        { "type": "boolean" },
        { "type": "string" }
      ],
      "unevaluatedItems": false,

      "$defs": {
        "closed": {
          "$anchor": "closed",
          "$ref": "#"
        }
      }
    }

Here the schema is "closed" to two array items. You can then later
use ``$ref`` and add another item like this:

.. schema_example::

    {
      "$id": "https://example.com/my-extended-tuple",

      "$ref": "https://example.com/my-tuple",
      "prefixItems": [
        { "type": "boolean" },
        { "type": "string" },
        { "type": "number" }
      ],
      "unevaluatedItems": false,

        "$defs": {

>           "extended": {

Reviewer: The intention here is a bit off (doesn't line up with the other properties).

Otherwise this looks reasonable to me I think at this point.

>         "$defs": {
>           "closed": {
>             "$anchor": "closed",
>             "$ref": "#",
>             "unevaluatedItems": false
>         }
>       }

Reviewer 2: ```suggestion
      "$defs": {
        "closed": {
          "$anchor": "closed",
          "$ref": "#",
          "unevaluatedItems": false
        }
      }
```

|draft2019-09|

The ``unevaluatedItems`` keyword applies to any values not evaluated
by an ``items``, ``prefixItems``, or ``contains`` keyword. Just as
``unevaluatedProperties`` affects only **properties** in an object,
``unevaluatedItems`` affects only **items** in an array.

.. note::
    Watch out! The word "unevaluated" *does not* mean "not evaluated by
    ``items``, ``prefixItems``, or ``contains``." "Unevaluated" means
    "not successfully evaluated", or "doesn't evaluate to true".

For this first example, let's make a "half-closed" schema: something
useful when you want to keep the first two arguments, but also add
more in certain situations. ("Closed" to two arguments in some
places, "open" to more arguments when you need it to be.)

.. schema_example::

    {
      "$id": "https://example.com/my-tuple",

      "type": "array",

>       "items": [

Author: I saw this! Wonder what happened. When did I change it to ``items``? Weird.

>         true,

Author: No. But I did think it was weird. Then again, { "type": "boolean" } looked weird alongside another { "type": "boolean" }. Wasn't entirely sure what to do here.

``unevaluatedItems`` like this:

.. schema_example::

    {
      "$id": "https://example.com/my-extended-tuple",

      "$ref": "https://example.com/my-tuple",
      "prefixItems": [
        true,
        true,
        { "type": "boolean" }
      ],
      "unevaluatedItems": false
    }

With this, you can use ``$ref`` to reference the first two
``prefixItems`` and keep the schema "closed" to two arguments when
you need it, "open" to more arguments when you need it. A reference to
``/my-tuple#closed`` would disallow more than two items. And because
``unevaluatedItems`` only sees inside its own subschema, if you
want to add an item, add it inside that subschema.

This means you can also put ``unevaluatedItems`` in a nested tuple.

.. schema_example::

    {
      "prefixItems": [
        { "type": "string" }
      ],
      "allOf": [
        {
          "prefixItems": [
            true,
            { "type": "number" }
          ]
        }
      ],
      "unevaluatedItems": false
    }
    --
    ["foo", 42]
    // All the array items are evaluated. The schema passes validation.
    --X
    ["foo", 42, null]
    // The third value is unevaluated, so ``unevaluatedItems`` is true,

>     // not false. The schema fails validation.

Reviewer 2: I'm not think this example says anything new that the previous examples didn't. It just inlines the reference. However, I like the way this is presented better with the passing and failing examples included.

> Note these two restraints, though. First, ``items`` doesn't "see

Reviewer 2: I feel like "restraints" is an awkward word to use here. "Restrictions" might make more sense, but I don't think that's quite right. I think it makes it sound like these are exceptions to the rule, which is not what's happening in these cases. I don't know what would be a better word to use.

inside" any instances of ``allOf``, ``anyOf``, or ``oneOf`` in the
same subschema. In this next example, ``items`` ignores ``allOf`` and
thus fails to validate.

.. schema_example::

    {

>       "allOf": [{ "prefixItems": [{ "enum": [1, "a"] }]  }],

Reviewer 2: ```suggestion
      "allOf": [{ "prefixItems": [{ "enum": [1, "a"] }] }],
```

      "items": { "const": 2 }
    }
    --X
    [1, "a", 2]

> 

Reviewer 2: This is a pretty important example. The `unevaluatedItems` keyword exists because this doesn't work. It might be a good idea to start with this (or a simpler version) followed by the same schema with `unevaluatedItems` in place of `items` as the solution to that problem.

Second, ``unevaluatedItems`` can't "see inside" cousins (vertically

> adjacent properties inside a separate pair of {curly braces} with the

Reviewer 2: ```suggestion
adjacent properties inside a separate subschema with the
```

same "parent"— ``anyOf``, ``if``, ``not``, or similar). For instance,
in the example below, the ``unevaluatedItems`` doesn't "see inside"
the ``prefixItems`` cousin before it. All instances fail validation

> because ``"prefixItems": [ true ]`` matches only length 1 arrays, and

Reviewer 2: ```suggestion
because ``"prefixItems": [true]`` matches only length 1 arrays, and
```

``{ "unevaluatedItems": false }`` matches only empty arrays.

.. schema_example::

    {
      "allOf": [
        { "prefixItems": [true] },
        { "unevaluatedItems": false }
      ]
    }
    --X

>     [1]

Reviewer 2: I'm not sure this needs to be included. It doesn't describe how the keyword works as much as how it doesn't work. If this was something that we frequently see people misunderstanding, it think it would be worth including, but I haven't seen that to be the case. I think this example can be left out.

.. note::
   For a tall list of more examples, read our `unevaluatedItems Test Suite <https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/unevaluatedItems.json>`_ on GitHub. We test a lot of use cases
   there, including uncommon ones.

Here are some of our examples in the suite:
   * ``unevaluatedItems`` nested inside another ``unevaluatedItems``
   * ``if/then/else`` statements interacting with ``unevaluatedItems``
   * nested ``if/then/else`` statements interacting with ``unevaluatedItems``
   * ``unevaluatedItems`` ignoring non-arrays
   * ``unevaluatedItems`` interacting with the ``not`` keyword

>    * and more

Reviewer 2: I'm not sure about the idea of linking to the test suite. My main concern is the links going out of date as the specification and the tests suite evolves. Another concern is consistency, we don't do that anywhere else in UJS. The other concern is that the audience of the test suite is very different that UJS. The audience of UJS is schema authors learning JSON Schema. The audience of the test suite is implementers. UJS shows how and when you might want to use a feature. The test suite is trying to break implementations.

The cross-linking could be useful, but I think I'm leaning toward preferring that it's not included.

|draft2019-09|

The ``unevaluatedItems`` keyword applies to any values not evaluated
by an ``items``, ``prefixItems``, or ``contains`` keyword. Just as
``unevaluatedProperties`` affects only **properties** in an object,
``unevaluatedItems`` affects only **item**-related keywords.

For this first example, let's assume you want to allow lists of items
that begin with either ``1`` or ``"A"``, but anything after must be ``2``.

.. schema_example::

    {
      "oneOf": [{"prefixItems": [{"const": 1}]}, {"prefixItems": [{"const": "a"}]}],
      "items": {"const": 2}

>     }

Author: Hmm, do you think it's important to mention somewhere that (line 257) ``items`` doesn't "see inside" any instances of ``oneOf``, ``anyOf``, or ``allOf`` in the same subschema? I mention in line 323 that ``unevaluatedItems`` doesn't "see inside" cousins, but it's difficult to weave those two threads together. Otherwise, I may scrap this entire example because you're right, it's pretty much the same as extending the tuple below.

>     }

Reviewer 2: Yes, I think it's important to mention because that's the reason `unevaluatedItems` exists. Going with my suggestion of starting with a simple open tuple and showing various ways of referencing it to extend or close it, you could start by showing how trying to extend or close  the referenced tuple with `items` doesn't work. Then you can introduce `unevaluatedItems` as the solution to that problem.

|draft2019-09|

> The ``unevaluatedItems`` keyword selects any data types not evaluated

Reviewer: (Keeping in mind for this and any other comments that I'm honestly not personally too familiar with conventions used throughout the book here, so if you see something I say contradicted somewhere feel free to say so!)

"values" is more correct than "data types" -- and instead of using the verb "selects" I'd probably use "applies validation to" or simply "applies to".

by an ``items``, ``prefixItems``, or `contains` keyword. Just as

> unevaluated"properties" affect only "properties" in an object, only

Reviewer: Bit more common I think to use emphasis -- e.g. unevaluated*Properties* -- rather than quotes.

"item"-related keywords affect unevaluated"items".

> It also applies inside valid subschemas with these keywords:

Reviewer: Ideally here we'd just link to another document which explains "applicator keywords" rather than repeating this but ok there's probably no other place you can link to here right now so I suppose it's fine.

- ``allOf``
- ``anyOf``
- ``oneOf``
- ``not``
- ``if``
- ``then``
- ``else``

The main reason to use this keyword is to extend an array with extra
arguments.

> For this first example, we'll use ``unevaluatedItems`` to select any

Author: I've edited all the other things on this page, but could you clarify what you mean here? My understanding is that ``unevaluatedItems`` works like ``unevaluatedProperties`` in that it _does_ recognize items declared in subschemas, whereas ``additionalProperties`` fails validation when attempting to extend a schema using ``allOf``. (I got this from the "Object" page.) For instance, I made this, which I validated with the schema validator:

```
{
  "allOf": [
    {
      "type": "array",
      "items": {
        "quantity": "number",
        "product": "string",
      },
    }
  ],

  "items": {
    "quantity": { "enum": ["1", "2", "3"] }
  },
    "required": ["quantity"],
  "unevaluatedItems": false
}
```
Or did I misunderstand something? You're the expert.

> For this first example, we'll use ``unevaluatedItems`` to select any

Reviewer: > unevaluatedItems works like unevaluatedProperties in that it does recognize items declared in subschemas, whereas additionalProperties fails validation when attempting to extend a schema using allOf

That's exactly right -- and your example I *think* maybe is fine, but what instances (data) were you going to use for that?

--

Regardless, what I meant is something like the following (sorry for being terse both previously and possibly still :)).

Imagine you have lists of numbers like `[1, 2, 2, 2, 2]`, `[1, 2, 2]`, `[1, 2]` -- so, a list where the first number is 1 and then you are allowed any number of elements after but they all have to be the number `2`.

So you write this schema:

```json
{
    "prefixItems": [{"const": 1}],
    "items": {"const": 2},
}
```

great, it works, lists that look like the above are valid, other lists are not.

Now you actually want to allow lists either like the ones above, or ones where the first number is `"a"`. OK great, let's modify the schema to also allow `"a"`

```json
{
    "oneOf": [{"prefixItems": [{"const": 1}]}, {"prefixItems": [{"const": "a"}]}],
    "items": {"const": 2},
}
```

because that's what `oneOf` does, right? Except no, that doesn't work! The reason is that `items` will now apply to *all* of the elements, requiring everything to be `2`. And indeed as you say `prefixItems` is now somehow not "hiding" elements from `items` the way it did in the simplest example.

However, `unevaluatedItems` indeed does what we want! If we just change `items` to `unevaluatedItems`, our schema does what we want -- lists starting with `[1, 2, ...]` are valid as before, as are `["a", 2, 2, ...]`, and others are invalid.

Now, my example is slightly contrived because you could also write the schema as:

```json
{
    "prefixItems": [{"oneOf": [{"const": 1}]}, {"const": "a"}]}],
    "items": {"const": 2},
}
```

and this would indeed work even with `items`, but I think this kind of example is clearer to start with.

Does that help at least explain what I mean (even if we don't necessarily change anything and stick with what you have)?

> For this first example, we'll use ``unevaluatedItems`` to select any

Author: Ah, okay. So the ``oneOf`` (I assume ``anyOf`` and ``allOf`` work the same way) is saying "one of either 1, a, or 2" BUT ALSO all items must be 2, therefore it fails if anything is not a 2. The ``xxxOf`` attaches itself to the ``items``, but ``unevaluatedItems`` are not evaluated by the ``xxxOf``.

I get it now, though this is an entirely different direction than I went. I did need an example of when ``uneavluatedItems`` is something other than a Boolean though, so I'll incorporate this. It's going to be a lot of examples under one subheader.

                "valid": false
            }
        ]
    }

Finally, here's an example of ``unevaluatedItems`` if you're
`structuring`. Let's make a "half-closed" schema: something useful
when you want to keep the first two arguments, but also add more in
certain situations. ("Closed" to two in some places, "open" to more
in others.)

.. schema_example::

    {
      "$id": "https://example.com/my-tuple",

      "type": "array",
      "prefixItems": [
        true,
        { "type": "boolean" }
      ],

      "$defs": {
        "closed": {
          "$anchor": "closed",
          "$ref": "#",
          "unevaluatedItems": false
        }
      }
    }

Then we can extend the tuple:

.. schema_example::

    {
      "$id": "https://example.com/my-extended-tuple",

      "$ref": "/my-tuple",
      "prefixItems": [
        true,
        true,
        { "type": "boolean" }
      ],
      "unevaluatedItems": false
    }

With this, you can use ``$ref`` to reference the first two
``prefixItems`` and keep the schema "closed" to two when necessary,
"open" to more when necessary. A reference to ``/my-tuple#closed``
would disallow more than two items, when you need it to.

.. note::

>    For a tower of more examples, read our `unevaluatedItems Test Suite <https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft2020-12/unevaluatedItems.json>`_ on GitHub.

Reviewer: "tower" is slightly awkward to me here, I'd say "long list" or if you want to be colorful, "torrent" I guess.

Author: Visual words help ideas stick better than abstract words, but sometimes I get too literary. I've been called out on it before by my writing instructor. I agree, "tower" is like a rock in the flow of the sentence.

   We test a lot of use cases there, including uncommon ones. Do any
   of these apply to your schema?

>    - using ``unevaluatedItems`` as a schema

Reviewer: I assume these were TODO notes to yourself perhaps from here down and meant to be removed?

Author: I tried to explain the base concepts of ``unevaluatedItems`` on this page and give jumping-off points for other cases readers might have. While I could have explained all the cases in the test suite, I felt five schema examples for one subheader was getting bloated. So I whittled down what I thought were the most abstract, general cases that needed explanation, then said essentially "Through this intro, I hope I've given you the tools to click to the test suite and understand some of the more uncommon, harder use cases, if they apply to you."

Reviewer: Ohhh I see, the bulleted list was meant to be the thing referred to by "Do any of these apply to your schema?" -- I have to say that wasn't even clear to me -- I though that sentence was trying to say "those" and simply saying "go check the test suite and see if any further examples apply to you".

I'm certainly supportive of not defining all the examples in this doc! Though I think we should then massage the bullets here, I have to reread them to see whether I can visualize which cases they're talking about. Thanks for the clarification.

   - ``unevaluatedItems`` nested inside another ``unevaluatedItems``
   - ``if/then/else`` statements with ``unevaluatedItems``
   - multiples nested ``if``/``then``s
   - multiple nested instances of ``contains``
   - ignoring non-array types

>    - ``not``

Reviewer: (Like `not` as a bullet I don't know what that refers to in context.)

|draft2019-09|

The ``unevaluatedItems`` keyword applies to any values not evaluated

> by an ``items``, ``prefixItems``, or `contains` keyword. Just as

Author: I was unsure whether to make this one a link to the ``contains`` subheader or not. After all, even though it comes immediately after the ``unevaluatedItems`` subheader, if a person is reading in order from the start of the site, they won't have read the ``contains`` section yet. So, ``contains`` is fine to not be a link? (single backticks)

> by an ``items``, ``prefixItems``, or `contains` keyword. Just as

Reviewer 2: Making it a link is fine, but the way this renders is that it uses the section header "Contains" rather as a keyword `contains`. There's a way to customize the text for the link, but I don't remember of the top of my head.

Also, I think it should be consistent. Either all the keywords link somewhere or none of them should.

``unevaluatedProperties`` affects only **properties** in an object,
``unevaluatedItems`` affects only **item**-related keywords.

For this first example, let's assume you want to allow lists of items
that begin with either ``1`` or ``"A"``, but anything after must be ``2``.

..schema_example::

{
    "oneOf": [{"prefixItems": [{"const": 1}]}, {"prefixItems": [{"const": "a"}]}],
    "items": {"const": 2},
}

The logic here seems like it should be "one of either ``1`` or ``"A"``
and then ``2``." Actually, it's "either ``1`` or ``"A"`` *and also* ``2``."
And ``items`` expects a ``2`` here, so anything that's not a ``2``
fails validation. That's because ``items`` doesn't "see inside" any
instances of ``oneOf``, ``anyOf``, or ``allOf`` in the same subschema.

But if you replace ``items`` with ``unevaluatedItems``, it passes.
Anything that starts with ``1`` or ``A`` and then continues with a ``2``
is valid in that case.

Because booleans are valid schemas for any JSON Schema keyword, you can
also prevent additional items by setting ``unevaluatedItems`` to
``false.`` In the next example, let's use ``unevaluatedItems`` to make
sure we allow no properties besides ``SKU`` and ``product``.

> .. note::
> Watch out! The word "unevaluated" *does not* mean "not evaluated by
> ``items``, ``prefixItems``, or ``contains``." "Unevaluated" means
> "not successfully evaluated", or "doesn't evaluate to true".

Reviewer 2: The note block needs to be indented in order to be rendered properly.

```suggestion
.. note::
    Watch out! The word "unevaluated" *does not* mean "not evaluated by
    ``items``, ``prefixItems``, or ``contains``." "Unevaluated" means
    "not successfully evaluated", or "doesn't evaluate to true".
```

("Closed" to two arguments in some places, "open" to more arguments
when you need it to be.)

.. schema_example::

    {
      "$id": "https://example.com/my-tuple",

      "type": "array",
      "prefixItems": [
        true,
        { "type": "boolean" }
      ],

      "$defs": {
        "closed": {
          "$anchor": "closed",
          "$ref": "#",
          "unevaluatedItems": false
        }
      }
    }

Then we can extend the tuple with another value:

.. schema_example::

    {
      "$id": "https://example.com/my-extended-tuple",

      "$ref": "https://example.com/my-tuple",
      "prefixItems": [
        true,
        true,
        { "type": "boolean" }
      ],
      "unevaluatedItems": false
    }

With this, you can use ``$ref`` to reference the first two
``prefixItems`` and keep the schema "closed" to two arguments when
you need it, "open" to more arguments when you need it. A reference to
``/my-tuple#closed`` would disallow more than two items.
``unevaluatedItems`` only sees inside its own subschema, so if you
want to add an item, add it inside that subschema.

This means you can also put ``unevaluatedItems`` in a nested tuple.

.. schema_example::

>     {
>         "schema": {
>             "prefixItems": [
>                 { "type": "string" }
>             ],
>             "allOf": [
>                 {
>                     "prefixItems": [
>                       true,
>                       { "type": "number" }
>                     ]
>                 }
>             ],
>             "unevaluatedItems": false
>         },
>         "tests": [
>             {
>                 "description": "with no unevaluated items",
>                 "data": ["foo", 42],
>                 "valid": true
>             },
>             {
>                 "description": "with an unevaluated item",
>                 "data": ["foo", 42, null],
>                 "valid": false
>             }
>         ]
>     }

Reviewer 2: Again, let's make use of the schema_example block features.

```suggestion
    {
      "prefixItems": [
        { "type": "string" }
      ],
      "allOf": [
        {
          "prefixItems": [
            true,
            { "type": "number" }
          ]
        }
      ],
      "unevaluatedItems": false
    }
    --
    ["foo", 42]
    --X
    ["foo", 42, null]
```

|draft2019-09|

The ``unevaluatedItems`` keyword applies to any values not evaluated
by an ``items``, ``prefixItems``, or `contains` keyword. Just as

> unevaluated**properties** affect only **properties** in an object, only

Author: It also looks like ReST doesn't let me bold only certain characters in a word, like every other markup language I know.

**item**-related keywords affect unevaluated**items**.

.. note::
Watch out! The word "unevaluated" *does not* mean "not evaluated by
``items``, ``prefixItems``, or ``contains``." "Unevaluated" means
"not successfully evaluated", or "doesn't evaluate to true".

For this first example, let's assume you want to allow lists of items
that begin with either "1" or "A," but anything after must be 2.

..schema_example::

{
    "oneOf": [{"prefixItems": [{"const": 1}]}, {"prefixItems": [{"const": "a"}]}],
    "items": {"const": 2},
}

The logic here seems like it should be "one of either 1 or A and then 2."
Actually, it's "either 1 or A **or 2** and also 2." So anything that's
not a 2 fails validation. That's because ``allOf``, ``anyOf``, and ``oneOf``
attach themselves to ``items`` in the same subschema.

But if you replace ``items`` with ``unevaluatedItems``, it passes.
Anything that starts with "1" or "A" and then continues with a "2" is
valid in that case. ``allOf``, ``anyOf``, and ``oneOf`` do not attach
themselves to ``unevaluatedItems``.

You can also define ``unevaluatedItems`` as a boolean. In the next

> example, let's use ``unevaluatedItems`` to make sure we have no values

Reviewer: ```suggestion
example, let's use ``unevaluatedItems`` to make sure we have no properties
```

outside of "SKU" and "product."

..schema_example::

    {
      "allOf": [
        {
          "type": "array",
          "items": {
            "SKU": "number",
            "product": "string",
          },
          "unevaluatedItems": false
        }
      ],

      "items": {
        "quantity": { "enum": ["1", "2", "3"] }
      },
        "required": ["quantity"],
    }

This schema will always fail validation because "quantity" is required,
but it's outside of the ``allOf``, and ``unevaluatedItems`` *does not recognize any values outside of its own subschema*.
Here, ``unevaluatedItems`` considers anything outside of "SKU" and
"product" to be additional. Combining the schemas with ``allOf`` won’t
change that.

> Instead, keep all your ``unevaluatedItems`` in the same list:

Author: "List of items" is what I mean, but "subschema" also works. "Items" just made me think of "list items" or <li>.

you need it, "open" to more arguments when you need it. A reference to
``/my-tuple#closed`` would disallow more than two items.
``unevaluatedItems`` only reads inside the subschema it's attached to,
so if you want to add an item, add it inside that subschema.

You can also set ``unevaluatedItems`` in a nested tuple.

.. schema_example::

    {
        "description": "unevaluatedItems with nested tuple",
        "schema": {
            "prefixItems": [
                { "type": "string" }
            ],
            "allOf": [
                {
                    "prefixItems": [
                      true,
                      { "type": "number" }
                    ]
                }
            ],
            "unevaluatedItems": false
        },
        "tests": [
            {
                "description": "with no unevaluated items",
                "data": ["foo", 42],
                "valid": true
            },
            {
                "description": "with unevaluated items",
                "data": ["foo", 42, null],
                "valid": false
            }
        ]
    }

In the first test, all the "data" values are evaluated, but in the

> second test, the ``null`` value is a type not specified by

Author: Yes. I'll make it clearer.
